### PR TITLE
Implement LSP debug test commands

### PR DESCRIPTION
**Pull request by dibarbet**
Implements server side of test debugging.  Part of https://github.com/dotnet/vscode-csharp/issues/5719

See the client PR for a gif - https://github.com/dotnet/vscode-csharp/pull/5965

### CODE REVIEW

File: src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LanguageServerResources.resx [hunk 1]
```diff
@@ -120,12 +120,24 @@
   <data name="Aborted" xml:space="preserve">
     <value>Aborted!</value>
   </data>
+  <data name="Attaching_debugger_to_process" xml:space="preserve">
```
Comment (Cosifne) [inline src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LanguageServerResources.resx:123]: It looks like should be called 
`Attaching_debugger_to_process_0`

File: src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Testing/TestRunner.cs [hunk 2]
```diff
@@ -35,10 +37,29 @@ internal partial class TestRunner(ILoggerFactory loggerFactory)
...
+        BufferedProgress<RunTestsPartialResult> progress,
+        VsTestConsoleWrapper vsTestConsoleWrapper,
+        TestRunHandler handler,
+        bool attachDebugger,
+        IClientLanguageServerManager clientLanguageServerManager)
+    {
+        if (attachDebugger)
+        {
+            // When we want to debug tests we need to use a custom test launcher so that we get called back with the process to attach to.
+            vsTestConsoleWrapper.RunTestsWithCustomTestHost(testCases, runSettings: null, handler, new DebugTestHostLauncher(progress, clientLanguageServerManager));
```
Comment (Cosifne) [inline src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Testing/TestRunner.cs:56]: Just to check, O# also doesn't support `runSettings`, is that correct?

File: src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Testing/TestRunner.DebugTestHostLauncher.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        }
+
+        private bool AttachDebugger(int processId, CancellationToken cancellationToken)
+        {
+            progress.Report(new RunTestsPartialResult(LanguageServerResources.Debugging_tests, string.Format(LanguageServerResources.Attaching_debugger_to_process, processId), Progress: null));
+
+            // Send an explicit request to the client to tell it to attach to the debugger and wait for the response.
+            // We want to wait for the attach to complete before we continue.
+            var task = Task.Run(async () => await AttachDebuggerAsync(processId, cancellationToken), cancellationToken);
+            return task.WaitAndGetResult_CanCallOnBackground(cancellationToken);
```
Comment (Cosifne) [inline src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Testing/TestRunner.DebugTestHostLauncher.cs:53]: I haven't looked at the client-side PR. But it feels like 
the client-side debug test command is waiting for the server debug test, and the server then waits for the client to attach the debugger to the process. 🤔 Are the 'debugger attaching' work lives in a separate thread?

Comment (Cosifne) [reply src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Testing/TestRunner.DebugTestHostLauncher.cs:53]: I just feel this looks like a deadlock, maybe it's fine as long as you test that works😄

**Comment by Cosifne** [comment]
A few minor questions and I will sign this off after they get answer :)

File: src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Testing/TestRunner.DebugTestHostLauncher.cs (hunk 3 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Testing/TestRunner.DebugTestHostLauncher.cs:53]: Not sure what you mean - but yes the client is not fully blocked waiting for the run tests command to complete.  It needs to be able to handle the partial results and this debug attach request while its waiting for the overall request to complete.

**Review by Cosifne — APPROVED** [review]

**Pull request merged**